### PR TITLE
simplifications + additions

### PR DESCRIPTION
**Pull request by liambuchanan**
simplifications:
* removes support for namespaces
* remove support for namespaceOverrides
* remove non standard defaults
* do not enforce specification of namespace

additions:
* add support for default overrides
* support for json schema draft 06
* exhaustive property tests
* support for additional labels

address open issues:
* [k8s 1.8+ support](https://github.com/better/kronjob/issues/12)
* [prefix container name with the job name](https://github.com/better/kronjob/issues/5)
* [include labels in jobTemplate metadata](https://github.com/better/kronjob/issues/3)

### CODE REVIEW

**Comment by imsky** [comment]
> removes support for namespaces
> remove support for namespaceOverrides

what does this mean for manifests using those features?

**Comment by liambuchanan** [comment]
we don't, but assuming someone somewhere does, this is a breaking change

**Comment by imsky** [comment]
datalake-etl makes use of namespaceOverrides and a bunch of jobs use namespaces though

**Comment by liambuchanan** [comment]
the datalake-etl jobs only exist in one env so they don't need namespaceOverrides (I updated that in a PR I'm just about to merge) and it's the list of namespaces that's being removed (not the ability to specify a namespace at all) which we don't use anywhere

**Comment by imsky** [comment]
hmm why remove namespaces though? what if we want to run an identical jobspec across multiple namespaces? is that still possible and, if so, how?

**Comment by liambuchanan** [comment]
@imsky yeah, just use jobs:

```
name: test
image: scratch
schedule: '* * * * *'
jobs:
  - namespace: prod
  - namespace: staging
```

**Comment by imsky** [comment]
hm ok, so it requires the jobs key. seems somewhat awkward to me - not obvious why jobs is required to specify namespaces.

```yaml
name: test
image: scratch
schedule: '* * * * *'
namespaces: [prod, staging]
```

seems cleaner IMO

**Comment by liambuchanan** [comment]
I agree that's marginally easier to read but I think the simplicity of only having a single metaproperty is worthwhile. you might also want identical jobs that pass different args to a command or use a different volume but adding specific syntax adds complexity.

**Comment by imsky** [comment]
hm i think specific change is is a bit of a downgrade :\

File: kronjob/_kronjob.py [hunk 1]
```diff
@@ -102,6 +102,7 @@ def serialize_k8s(k8s_object):
 
 
 def build_k8s_object(aggregate_job):
+    _ALTERNATE_DEFAULTS['containerName'] = '{}-job'.format(aggregate_job['name'])
```
Comment (imsky) [inline kronjob/_kronjob.py]: should check to make sure this is under 63 chars in length, think kubernetes has a limit on container names, and there are other issues (see https://github.com/better/mortgage/pull/18124)

probably should make the max length configurable

File: kronjob/_kronjob.py [hunk 2]
```diff
@@ -136,6 +136,7 @@ def _get_args(*keys):
             metadata=metadata,
             spec=k8s_models.V2alpha1CronJobSpec(
                 job_template=k8s_models.V2alpha1JobTemplateSpec(
+                    metadata=k8s_models.V1ObjectMeta(labels=labels),
```
Comment (imsky) [inline kronjob/_kronjob.py]: btw if you add `fix #3` to the commit message, it'll close the original issue

File: kronjob/_kronjob.py [hunk 3]
```diff
@@ -101,29 +78,62 @@ def serialize_k8s(k8s_object):
...
+        JobTemplateSpec = k8s_models.V1beta1JobTemplateSpec
+    else:
+        cronjob_api_version = 'v2alpha1'
+        CronJob = k8s_models.V2alpha1CronJob
+        CronJobSpec = k8s_models.V2alpha1CronJobSpec
+        JobTemplateSpec = k8s_models.V2alpha1JobTemplateSpec
+
+    defaults = copy.deepcopy(defaults) if defaults is not None else {}
+    if 'containerName' not in defaults:
+        defaults['containerName'] = '{}-job'.format(aggregate_job['name'])
```
Comment (imsky) [inline kronjob/_kronjob.py:102]: should check to make sure this is under 63 chars in length, think kubernetes has a limit on container names, and there are other issues (see https://github.com/better/mortgage/pull/18124)

probably should make the max length configurable

File: kronjob/_kronjob.py [hunk 4]
```diff
@@ -101,29 +78,62 @@ def serialize_k8s(k8s_object):
...
         }
 
-    labels = {aggregate_job.get('labelKey', 'kronjob/job'): aggregate_job['name']}
-    metadata = k8s_models.V1ObjectMeta(
-        labels=labels,
-        **_get_args('name', 'namespace')
-    )
+    labels = _get_arg('labels')
+    labels[_get_arg('labelKey')] = _get_arg('name')
+    metadata = k8s_models.V1ObjectMeta(labels=labels, **_get_args('name', 'namespace'))
```
Comment (imsky) [inline kronjob/_kronjob.py:119]: btw if you add `fix #3` to the commit message, it'll close the original issue

File: kronjob/_kronjob.py (hunk 3 above)
Comment (liambuchanan) [reply kronjob/_kronjob.py:102]: 👍 I'll add a validation. I'll probably just hard code 63 chars for now unless the limit's likely to change for some reason?

File: kronjob/_kronjob.py (hunk 4 above)
Comment (liambuchanan) [reply kronjob/_kronjob.py:119]: oh cool. the more you know

File: kronjob/_kronjob.py (hunk 3 above)
Comment (liambuchanan) [reply kronjob/_kronjob.py:102]: I added a validation on name since it has to be <=52 chars for cronjobs which will act as a proxy for this.

**Comment by imsky** [comment]
before this is merged and published, would be good to pin the version of kronjob we install in scripts

should probably release this early next week

**Comment by imsky** [comment]
👍

**Comment by liambuchanan** [comment]
thanks @imsky. everything should be pinned atm (https://github.com/better/mortgage/pull/18752) will hold off on release till monday

**Pull request merged**